### ISSUE TITLE
Options with multiple typed params (was: Unmatched arguments when using multiple groups)

### ISSUE THREAD

**Issue by melsabagh-kw** [remkop/picocli#1283]
I am expecting Picocli to be able to successfully parse arguments matching the synopsis: `cmd (<a1> <a2>) (<b1> <b2>)`.  However, Picocli fails at parsing arguments beyond the `a2` parameter. Here is a test case that produces the issue: 

```java
public class MultigroupTest {

    static class GroupA {
        @CommandLine.Parameters
        String a1;
        @CommandLine.Parameters
        String a2;
    }

    static class GroupB {
        @CommandLine.Parameters
        String b1;
        @CommandLine.Parameters
        String b2;
    }

    @Test
    public void testMultigroups() {
        @CommandLine.Command(name = "cmd")
        class SomeCommand {
            @CommandLine.ArgGroup(exclusive = false, multiplicity = "1")
            GroupA groupA;

            @CommandLine.ArgGroup(exclusive = false, multiplicity = "1")
            GroupB groupB;
        }

        new CommandLine(new SomeCommand()).parseArgs("a1", "a2", "b1", "b2");
    }
}
```

Picocli currently fails the above test and throws the following exception:

```
picocli.CommandLine$UnmatchedArgumentException: Unmatched arguments from index 2: 'b1', 'b2'
        at picocli.CommandLine$Interpreter.validateConstraints(CommandLine.java:12439)
        at picocli.CommandLine$Interpreter.parse(CommandLine.java:12391)
        at picocli.CommandLine$Interpreter.parse(CommandLine.java:12259)
        at picocli.CommandLine.parseArgs(CommandLine.java:1458)
```

This issue seems to be present in the latest release and in `master`. The number of parameters in `GroupB` (the 2nd group) does not seem to matter. Replacing group `Parameters` with group `Option` triggers the same issue.

**Comment by remkop** [comment]
Hi @melsabagh-kw, I think you may have bumped into one of the [limitations of argument groups](https://picocli.info/#_argument_group_limitations): defining positional parameters in multiple groups does not work correctly.

This is to do with the way the index of these parameters is tracked. I am not sure if I will be able to improve this...

**Comment by melsabagh-kw** [comment]
Thanks @remkop for the prompt reply. Picocli is awesome! Could you elaborate a bit more on the underlying issue? I would appreciate any thoughts on potential workarounds.

**Comment by remkop** [comment]
The problem has to do with index tracking. GroupA has `@Parameter String a1` (implicit `index = "0"`) and `@Parameter String a2` (implicit `index = "1"`). Group B has a similar set of params, also with implicit index 0 and 1. There is no way for picocli to find out which group positional parameters should be assigned to.

For your particular example, I would suggest having a single group, and set multiplicity to 2, so that users must specify it twice.

```java
        class SomeCommand {
            @CommandLine.ArgGroup(exclusive = false, multiplicity = "2")
            GroupA groupA;
        }
```

Basically, positional parameters can only occur in one group.

PS. Glad you like picocli! 🙏  Please star ⭐ picocli on GitHub if you like the project! 😉

**Comment by remkop** [comment]
@melsabagh-kw Was that workaround useful for you?

**Comment by melsabagh-kw** [comment]
@remkop not really. My real use case involves groups of different cardinality (actually multi-arg options with different types/labels, which I was using groups to emulate). I have decided to just drop positional params altogether, and assign an option to each argument and let the caller deal with it :-). It would be nice if Picocli had support for multi-arg options with multiple param labels like Python's argparse.

**Comment by remkop** [comment]
@melsabagh-kw I see what you mean, kind of like [`@SubParameters` in JCommander](https://jcommander.org/#_with_a_class). I need to think about this more...

With the current version of picocli, you can achieve this (without a group) by using a [custom parameter consumer](https://picocli.info/#_custom_parameter_processing). For example:

```java
class Foo {
    String a;
    String b;

    Foo(String a, String b) { this.a = a; this.b = b; }
}

class Command {
    @Option(names = "--foo", description = "Foo's a and b", parameterConsumer = FooParameterConsumer.class)
    Foo foo;
}

class FooParameterConsumer implements IParameterConsumer {
    public void consumeParameters(Stack<String> args, ArgSpec argSpec, CommandSpec commandSpec) {
        if (args.size() < 2) {
            throw new ParameterException(commandSpec.commandLine(), "--foo requires two parameters");
        }
        String a = args.pop();
        String b = args.pop();
        argSpec.setValue(new Foo(a, b));
    }
}
```

**Comment by remkop** [comment]
@melsabagh-kw Does using a [custom parameter consumer](https://picocli.info/#_custom_parameter_processing) meet your needs?

I have thought about it a little, but it may be difficult to support options with multiple typed parameters with only annotations.

We may want something like this to work out of the box:

```java
class Fraction {
  @Parameter(index = "0") int numerator;
  @Parameter(index = "1") int denominator;
}

class Cmd {
    @Option(names = "--fraction", arity = "2")
    Fraction f;
}
```

This would require fairly large changes to picocli:
* having an `arity` greater than 1 (one) currently only works for multi-value types like arrays, Collections and Maps. Picocli would need to somehow recognize custom types as multi-value. One way to do that would be implementing a `Collection` interface like `java.util.List`, which would be a lot of work for the implementor. Another way would be to inspect custom types for the presence of `@Parameter` annotated fields and methods, and match the arity of the option with the specified  `@Parameter` index ranges.
* the [type converter](https://picocli.info/#_custom_type_converters) interface only processes one parameter at a time; so the parser would process option parameters one at a time, creating instances and setting `@Parameter` annotated fields one by one. Probably doable but this would add quite a bit of book-keeping to the parser logic.

The simplest thing is to use a [custom parameter consumer](https://picocli.info/#_custom_parameter_processing) like I suggested in my previous comment. That is very similar to a custom type converter that can handle an arbitrary number of parameters.

Thoughts?

**Comment by remkop** [comment]
Closing this ticket. Please feel free to re-open or create another ticket if the suggestion to use a custom parameter consumer does not meet your requirements.

Enjoy picocli!

**Comment by melsabagh-kw** [comment]
@remkop sorry I just saw your last two responses. A custom parameter consumer technically can be used to do what I want, but I hate the extra code bloat.

**Comment by remkop** [comment]
It is always a difficult decision where to draw the line between which functionality should go into the library or into the application. My rule of thumb is that common use cases should be easy, while uncommon use cases should be possible. 

With the current version of picocli, there is a convenient `arity = "1..n"` annotation API for applications with options that take multiple parameters that have the same type. If option parameters need to have different types, I think this is rare enough that picocli should make this possible but perhaps not necessarily easy/convenient.

Note that there is also a different approach: applications can also require a single argument, that contains comma-separated values, so in the parameter label the application can say something like `paramLabel = "X,Y"`.
